ledgerManager.php

```diff
@@ -0,0 +1,319 @@
+	 * @param string $endDate
+	 */
+	public function getCreditSumForWindow($startDate, $endDate)
+	{
+		return $this->getTotalSumCreditTransactions(true, $startDate, $endDate);
+	}
+
+	/**
+	 * @param string $startDate
+	 * @param string $endDate
+	 */
+	public function getDebitSumForWindow($startDate, $endDate)
+	{
+		return $this->getTotalSumCreditTransactions(false, $startDate, $endDate);
+	}
+
+	/**
+	 * This method returns a balance for a given date, now() by default
+	 *
+	 * @param string|null $startDate
+	 * @param string|null $endDate
+	 * @return float
+	 */
+	public function getBalance($startDate = null, $endDate = null)
+	{
+		if ($startDate || $endDate) {
+			return $this->calculateBalance($startDate, $endDate);
+		} elseif (!isset($this->overallBalance)) {
+			return $this->calculateBalance();
+		}
+		return $this->overallBalance;
+	}
+
+	/**
+	 * @return float
+	 */
+	public function getUnclearedAmount()
+	{
+		$query = $this->pdo->query("
+			select sum(amount) amount
+			from {$this->table}
+			where credit = false and cleared = false
+		");
+		$row = $query->fetch(PDO::FETCH_ASSOC);
+		return (float)$row['amount'];
+	}
+
+	/**
+	 * @return int
+	 */
+	public function numberOfDaysLeftInPayPeriod()
+	{
+		$todaysDate = (int)date('d');
+		if ($todaysDate <= 15) {
+			return 15 - $todaysDate + 1;
+		} else {
+			$query = $this->pdo->query("
+				select (datediff(last_day(now()), now()) + 1) as daysLeft
+			");
+			$row = $query->fetch(PDO::FETCH_ASSOC);
+			return (int)$row['daysLeft'];
+		}
+	}
+
+	/**
+	 * This window retrieves an array of transactions for a given date range.
+	 * Note: The windowStartDate serves as a cutoff in terms of the returned date
+	 * BUT will not affect the rolling balance which is calculated through the
+	 * beginning of time.  ie. For a database containing transactions going back
+	 * to 2014, a windowStartDate of 2015 would return 2015 through now(), but
+	 * but the transactions in 2014 would setup the balance for 2015.
+	 *
+	 * @param string $windowStartDate
+	 * @param string|null $endDate
+	 * @param float|null $runningBalance
+	 * @return mixed[]
+	 */
+	public function retrieveARangeOfTransactions($windowStartDate, $endDate = null)
+	{
+		$allTransactions = [];
+		$runningBalance = $this->getBalance(null, $endDate);
+
+		$queryString = "
+			select id, credit, description, amount, time, cleared
+			from {$this->table}
+			where time >= :windowStartDate
+		";
+		if ($endDate) {
+			$queryString .= "and time <= :endDate ";
+		}
+		$queryString .= "order by time desc";
+		$query = $this->pdo->prepare($queryString);
+		$query->bindParam(':windowStartDate', $windowStartDate);
+		if ($endDate) {
+			$query->bindParam(':endDate', $endDate);
+		}
+		$query->execute();
+
+		$rows = $query->fetchAll(PDO::FETCH_ASSOC);
+		foreach ($rows as $row) {
+			$row['balance'] = $runningBalance;
+			$allTransactions[] = $row;
+			// Realize that we're calculating balances backwards in time
```

Comment:
"we have the current transaction's balance, so we need to ADD debits and SUB credits to calculate the next, previous-in-time balance"